### PR TITLE
Payment secret support

### PR DESCRIPTION
**Pull request by rustyrussell**
Based on #3167 this actually uses the onion to complete the payment_secret flow.

Changelog-None: EXPERIMENTAL only

### CODE REVIEW

**Comment by cdecker** [comment]
Rebased on top of `master` after merging #3167

**Comment by rustyrussell** [comment]
Trivial rebase to fix flake8 complaint.

File: devtools/bolt11-cli.c [hunk 1]
```diff
@@ -1,6 +1,7 @@
 #include <bitcoin/address.h>
 #include <bitcoin/base58.h>
 #include <bitcoin/chainparams.h>
+#include <bitcoin/privkey.h>
```
[the remark this answers is not in the document]
Comment (rustyrussell) [reply devtools/bolt11-cli.c]: struct secret is in there....

**Comment by rustyrussell** [comment]
Note: I'm deferring this until after #3278

**Comment by cdecker** [comment]
#3278 has been updated in order to address the feedback, so hopefully I'm not blocking this PR much longer ^^

**Comment by rustyrussell** [comment]
Rebased.  Lisa, since your change didn't get merged, does that mean this is still buggy?

**Comment by rustyrussell** [comment]
Ack https://github.com/ElementsProject/lightning/commit/613172be706b29ce14abfee80c90f977b8b00c5f

AFAICT we handle the length prefix properly:

```C
	/* Copy common pieces over */
	step->raw_payload = tal_dup_arr(step, u8, src, raw_size + vsize, 0);
	memcpy(step->next->mac, src + hop_size - HMAC_SIZE, HMAC_SIZE);

...
	else if (step->type == SPHINX_TLV_PAYLOAD) {
		const u8 *tlv = step->raw_payload;
		size_t max = tal_bytelen(tlv);
		step->payload.tlv = tlv_tlv_payload_new(step);

		/* The raw payload includes the length / realm prefix, Consume
		 * the length off of the payload, so the decoding can strat
		 * correctly. */
		fromwire_varint(&tlv, &max);

		if (!fromwire_tlv_payload(&tlv, &max, step->payload.tlv)) {
			/* FIXME: record offset of violation for error! */
			step->type = SPHINX_INVALID_PAYLOAD;
			return;
		}
	}
```

**Pull request merged**